File: src/importers.ts
import { normalizeAnthropicUsage } from "./adapters";
import { validateEvent, type EconomyEvent } from "./events";
import { createHash } from "node:crypto";

export interface ImportOptions { projectId: string; sourceId: string }
type Row = Record<string, unknown>;
type UsageEvent = Extract<EconomyEvent, { eventType: "usage" }>;
const record = (value: unknown): Row | null => value !== null && typeof value === "object" && !Array.isArray(value) ? value as Row : null;
const string = (value: unknown): string | null => typeof value === "string" && value.trim() ? value : null;
const count = (value: unknown): number | null => typeof value === "number" && Number.isSafeInteger(value) && value >= 0 ? value : null;
const stamp = (value: unknown): string | null => {
  if (typeof value !== "string") return null;
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? null : date.toISOString();
};
function rows(text: string): Row[] {
  return text.split(/\r?\n/).flatMap((line, index) => {
    if (!line.trim()) return [];
    try { const parsed = record(JSON.parse(line)); if (!parsed) throw new Error("not an object"); return [parsed]; }
    catch (error) { throw new Error(`Invalid JSONL line ${index + 1}: ${error instanceof Error ? error.message : "parse failed"}`); }
  });
}
function base(options: ImportOptions, sessionId: string, sourceEventId: string, timestamp: string, clientId: string, requestId: string | null, modelId: string | null): Omit<UsageEvent, "eventType" | "payload"> {
  if (!options.projectId.trim() || !options.sourceId.trim()) throw new Error("projectId and sourceId are required");
  const eventId = createHash("sha256").update(JSON.stringify([options.projectId, options.sourceId, sourceEventId])).digest("hex");
  return { schemaVersion: 1, eventId, occurredAt: timestamp,
    sourceId: options.sourceId, sourceEventId, projectId: options.projectId, taskId: null,
    sessionId, requestId, attemptId: null, clientId, adapterVersion: "experimental-jsonl-1",
    modelId, providerId: clientId === "claude" ? "anthropic" : "openai",
    executionLocation: "unknown", evidence: "provider_reported" };
}

/** Explicit, read-only import of Claude Code assistant transcript rows. */
export function importClaudeTranscript(text: string, options: ImportOptions): EconomyEvent[] {
  const selected = new Map<string, { row: Row; message: Row; usage: Row; session: string; id: string; time: string }>();
  for (const row of rows(text)) {
    if (row.type !== "assistant") continue;
    const message = record(row.message);
    const usage = record(message?.usage);
    const id = string(message?.id);
    const time = stamp(row.timestamp);
    if (!message || !usage || !id || !time) continue;
    const session = string(row.sessionId) ?? options.sourceId;
    const key = `${session}:${id}`;
    const prior = selected.get(key);
    if (!prior || time >= prior.time) selected.set(key, { row, message, usage, session, id, time });
  }
  return [...selected.values()].map(({ row, message, usage, session, id, time }) => {
    const normalized = normalizeAnthropicUsage(usage);
    return validateEvent({ ...base(options, session, `${session}:message:${id}`, time, "claude", string(row.requestId) ?? id, string(message.model)),
      eventType: "usage", payload: { inputTokens: normalized.inputTokens, outputTokens: normalized.outputTokens,
        cacheReadTokens: normalized.cacheReadTokens, cacheWriteTokens: normalized.cacheWriteTokens,
        complete: string(message.stop_reason) !== null, category: "primary", callId: `${session}:${id}`, costUsd: null,
        observationScope: "call", rawUsage: usage, normalizerVersion: "anthropic-1" } });
  });
}

type Counters = { input: number; output: number; read: number | null; write: number | null };
function codexCounters(value: unknown): Counters | null {
  const usage = record(value);
  const input = count(usage?.input_tokens), output = count(usage?.output_tokens);
  const read = count(usage?.cached_input_tokens), write = count(usage?.cache_write_input_tokens);
  return input !== null && output !== null ? { input, output, read, write } : null;
}
function delta(current: Counters, previous: Counters): Counters | null {
  if (current.input < previous.input || current.output < previous.output ||
    (current.read !== null && previous.read !== null && current.read < previous.read) ||
    (current.write !== null && previous.write !== null && current.write < previous.write)) return null;
  return { input: current.input - previous.input, output: current.output - previous.output,
    read: current.read === null || previous.read === null ? null : current.read - previous.read,
    write: current.write === null || previous.write === null ? null : current.write - previous.write };
}

/** Explicit, read-only import of Codex rollout cumulative token_count events. */
export function importCodexRollout(text: string, options: ImportOptions): EconomyEvent[] {
  let session = options.sourceId;
  let previous: Counters = { input: 0, output: 0, read: 0, write: 0 };
  let hasSessionStart = false;
  let hasBaseline = false;
  const result: EconomyEvent[] = [];
  for (const [index, row] of rows(text).entries()) {
    const payload = record(row.payload);
    if (row.type === "session_meta") {
      session = string(payload?.id) ?? session;
      previous = { input: 0, output: 0, read: 0, write: 0 };
      hasBaseline = false;
      hasSessionStart = true;
      continue;
    }
    if (row.type !== "event_msg" || payload?.type !== "token_count") continue;
    const info = record(payload.info);
    const current = codexCounters(info?.total_token_usage);
    const time = stamp(row.timestamp);
    if (!current || !time) continue;
    if (!hasBaseline) {
      hasBaseline = true;
      if (!hasSessionStart) { previous = current; continue; }
    }
    const change = delta(current, previous);
    if (!change) { previous = current; continue; }
    previous = current;
    if (Object.values(change).every(value => value === 0 || value === null)) continue;
    result.push(validateEvent({ ...base(options, session, `${session}:token_count:${index}`, time, "codex", null, null),
      eventType: "usage", payload: { inputTokens: change.input, outputTokens: change.output,
        cacheReadTokens: change.read, cacheWriteTokens: change.write,
        complete: false, category: "primary", callId: null, costUsd: null,
        observationScope: "session_delta", rawUsage: record(info?.total_token_usage) ?? {},
        normalizerVersion: "codex-cumulative-1" } }));
  }
  return result;
}

File: python/julius/importers.py
"""Explicit JSONL imports. No filesystem scanning or model execution."""

from __future__ import annotations

from datetime import datetime, timezone
import hashlib
import json
from typing import Any

from .adapters import normalize_anthropic_usage
from .events import validate_event


def _obj(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _string(value: Any) -> str | None:
    return value if isinstance(value, str) and value.strip() else None


def _count(value: Any) -> int | None:
    return value if type(value) is int and 0 <= value <= 2**53 - 1 else None


def _time(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            return None
        return (
            parsed.astimezone(timezone.utc)
            .isoformat(timespec="milliseconds")
            .replace("+00:00", "Z")
        )
    except ValueError:
        return None


def _rows(text: str) -> list[dict[str, Any]]:
    result = []
    for index, line in enumerate(text.splitlines(), 1):
        if not line.strip():
            continue
        try:
            value = json.loads(line)
            if not isinstance(value, dict):
                raise ValueError("not an object")
            result.append(value)
        except (ValueError, json.JSONDecodeError) as error:
            raise ValueError(f"Invalid JSONL line {index}: {error}") from error
    return result


def _base(
    options: dict[str, str],
    session: str,
    source_event: str,
    time: str,
    client: str,
    request: str | None,
    model: str | None,
) -> dict[str, Any]:
    project, source = options.get("projectId"), options.get("sourceId")
    if not _string(project) or not _string(source):
        raise ValueError("projectId and sourceId are required")
    event_id = hashlib.sha256(
        json.dumps([project, source, source_event], separators=(",", ":")).encode()
    ).hexdigest()
    return {
        "schemaVersion": 1,
        "eventId": event_id,
        "occurredAt": time,
        "sourceId": source,
        "sourceEventId": source_event,
        "projectId": project,
        "taskId": None,
        "sessionId": session,
        "requestId": request,
        "attemptId": None,
        "clientId": client,
        "adapterVersion": "experimental-jsonl-1",
        "modelId": model,
        "providerId": "anthropic" if client == "claude" else "openai",
        "executionLocation": "unknown",
        "evidence": "provider_reported",
        "eventType": "usage",
    }


def import_claude_transcript(text: str, options: dict[str, str]) -> list[dict[str, Any]]:
    selected: dict[str, tuple[dict[str, Any], dict[str, Any], dict[str, Any], str, str, str]] = {}
    for row in _rows(text):
        if row.get("type") != "assistant":
            continue
        message = _obj(row.get("message"))
        usage = _obj(message.get("usage"))
        message_id, time = _string(message.get("id")), _time(row.get("timestamp"))
        if not usage or not message_id or not time:
            continue
        session = _string(row.get("sessionId")) or options.get("sourceId", "")
        key = f"{session}:{message_id}"
        if key not in selected or time >= selected[key][5]:
            selected[key] = row, message, usage, session, message_id, time
    result = []
    for row, message, usage, session, message_id, time in selected.values():
        normalized = normalize_anthropic_usage(usage)
        event = _base(
            options,
            session,
            f"{session}:message:{message_id}",
            time,
            "claude",
            _string(row.get("requestId")) or message_id,
            _string(message.get("model")),
        )
        event["payload"] = {
            "inputTokens": normalized["inputTokens"],
            "outputTokens": normalized["outputTokens"],
            "cacheReadTokens": normalized["cacheReadTokens"],
            "cacheWriteTokens": normalized["cacheWriteTokens"],
            "complete": _string(message.get("stop_reason")) is not None,
            "category": "primary",
            "callId": f"{session}:{message_id}",
            "costUsd": None,
            "observationScope": "call",
            "rawUsage": usage,
            "normalizerVersion": "anthropic-1",
        }
        result.append(validate_event(event))
    return result


def _counters(value: Any) -> tuple[int, int, int | None, int | None] | None:
    usage = _obj(value)
    input_tokens, output = _count(usage.get("input_tokens")), _count(usage.get("output_tokens"))
    return (
        (
            input_tokens,
            output,
            _count(usage.get("cached_input_tokens")),
            _count(usage.get("cache_write_input_tokens")),
        )
        if input_tokens is not None and output is not None
        else None
    )


def import_codex_rollout(text: str, options: dict[str, str]) -> list[dict[str, Any]]:
    session = options.get("sourceId", "")
    previous: tuple[int, int, int | None, int | None] = (0, 0, 0, 0)
    has_session_start = has_baseline = False
    result = []
    for index, row in enumerate(_rows(text)):
        payload = _obj(row.get("payload"))
        if row.get("type") == "session_meta":
            session = _string(payload.get("id")) or session
            previous, has_baseline, has_session_start = (0, 0, 0, 0), False, True
            continue
        if row.get("type") != "event_msg" or payload.get("type") != "token_count":
            continue
        raw = _obj(_obj(payload.get("info")).get("total_token_usage"))
        current, time = _counters(raw), _time(row.get("timestamp"))
        if current is None or time is None:
            continue
        if not has_baseline:
            has_baseline = True
            if not has_session_start:
                previous = current
                continue
        if any(c is not None and p is not None and c < p for c, p in zip(current, previous)):
            previous = current
            continue
        change = tuple(
            c - p if c is not None and p is not None else None for c, p in zip(current, previous)
        )
        previous = current
        if all(value in (0, None) for value in change):
            continue
        event = _base(options, session, f"{session}:token_count:{index}", time, "codex", None, None)
        event["payload"] = {
            "inputTokens": change[0],
            "outputTokens": change[1],
            "cacheReadTokens": change[2],
            "cacheWriteTokens": change[3],
            "complete": False,
            "category": "primary",
            "callId": None,
            "costUsd": None,
            "observationScope": "session_delta",
            "rawUsage": raw,
            "normalizerVersion": "codex-cumulative-1",
        }
        result.append(validate_event(event))
    return result
